File: sdk/swift/Sources/SeocheonSDK/Constants/ChainConstants.swift
import Foundation

/// Chain parameter constants for the Seocheon blockchain.
public enum ChainConstants {
    // MARK: - Epoch/Window parameters

    /// Blocks per epoch (~24h at 5s/block)
    public static let epochLength: Int64 = 17280
    /// Windows per epoch
    public static let windowsPerEpoch: Int64 = 12
    /// Minimum active windows for reward qualification
    public static let minActiveWindows: Int64 = 8
    /// Blocks per window (epochLength / windowsPerEpoch)
    public static let windowLength: Int64 = 1440

    // MARK: - Quota parameters

    /// Epoch quota for self-funded nodes
    public static let selfFundedQuota: UInt64 = 100
    /// Epoch quota for feegrant nodes
    public static let feegrantQuota: UInt64 = 10

    // MARK: - Pruning parameters

    /// Activity pruning keep blocks (~1 year)
    public static let activityPruningKeepBlocks: Int64 = 6307200

    // MARK: - Fee model parameters

    public static let feeThresholdMultiplier: Int64 = 3
    /// 1 KKOT in uppyeo
    public static let baseActivityFee: Int64 = 10_000_000_000
    /// Basis points (0.5)
    public static let feeExponent: Int64 = 5000
    /// 100 KKOT in uppyeo
    public static let maxActivityFee: Int64 = 1_000_000_000_000
    public static let minFeegrantQuota: UInt64 = 8
    /// Basis points (0.5)
    public static let quotaReductionRate: Int64 = 5000
    public static let feegrantFeeExempt: Bool = true

    // MARK: - Dual reward pool parameters

    /// D_min basis points (0.3)
    public static let dMin: Int64 = 3000
    /// Fee to activity pool ratio basis points (0.8)
    public static let feeToActivityPoolRatio: Int64 = 8000

    // MARK: - Node registration parameters

    public static let maxRegistrationsPerBlock: Int64 = 5
    public static let registrationCooldownBlks: Int64 = 100
    public static let registrationDeposit: String = "0"
    public static let maxTags: Int64 = 10
    public static let maxTagLength: Int64 = 32

    // MARK: - Agent permission parameters

    public static let agentAllowedMsgTypes: [String] = [
        "/seocheon.activity.v1.MsgSubmitActivity",
        "/cosmos.bank.v1beta1.MsgSend",
    ]
    public static let agentFeegrantAllowedMsgTypes: [String] = [
        "/seocheon.activity.v1.MsgSubmitActivity",
    ]
    public static let agentAddressChangeCooldown: Int64 = 17280

    // MARK: - Time-block conversion (5s/block)

    public static let blocksPerHour: Int64 = 720
    public static let blocksPerDay: Int64 = 17280
    public static let blocksPerYear: Int64 = 6307200
    /// ~21 days
    public static let unbondingPeriodBlocks: Int64 = 362880
    /// ~180 days
    public static let feegrantExpiryBlocks: Int64 = 3110400

    // MARK: - Token denomination constants (6-stage system)

    /// Base denomination (10^0)
    public static let tokenBaseDenom: String = "uppyeo"
    /// sal denomination (10^2)
    public static let tokenSalDenom: String = "sal"
    /// pi denomination (10^4)
    public static let tokenPiDenom: String = "pi"
    /// sum denomination (10^6)
    public static let tokenSumDenom: String = "sum"
    /// hon denomination (10^8)
    public static let tokenHonDenom: String = "hon"
    /// Display denomination (10^10)
    public static let tokenDisplayDenom: String = "kkot"

    // MARK: - Denomination conversion factors

    public static let uppyeoPerSal: Int64 = 100
    public static let uppyeoPerPi: Int64 = 10_000
    public static let uppyeoPerSum: Int64 = 1_000_000
    public static let uppyeoPerHon: Int64 = 100_000_000
    public static let uppyeoPerKkot: Int64 = 10_000_000_000

    // MARK: - Default SDK configuration

    public static let defaultGasPrice: String = "250uppyeo"
    public static let defaultGasAdjustment: Double = 1.3
    public static let defaultBroadcastMode: String = "sync"
    public static let defaultConfirmTimeoutMs: UInt64 = 30000
    public static let defaultConfirmPollMs: UInt64 = 1000
}

File: sdk/kotlin/src/main/kotlin/com/seocheon/sdk/constants/ChainConstants.kt
package com.seocheon.sdk.constants

/**
 * Chain parameters, token denominations, and default SDK configuration values
 * for the Seocheon blockchain.
 */
object ChainConstants {

    // Epoch/Window parameters (x/activity params)
    const val EPOCH_LENGTH: Long = 17280          // blocks per epoch (~24h at 5s/block)
    const val WINDOWS_PER_EPOCH: Long = 12        // windows per epoch
    const val MIN_ACTIVE_WINDOWS: Long = 8        // minimum active windows for reward qualification
    const val WINDOW_LENGTH: Long = 1440           // blocks per window (EPOCH_LENGTH / WINDOWS_PER_EPOCH)

    // Quota parameters (x/activity params)
    const val SELF_FUNDED_QUOTA: Long = 100        // epoch quota for self-funded nodes
    const val FEEGRANT_QUOTA: Long = 10            // epoch quota for feegrant nodes

    // Pruning parameters
    const val ACTIVITY_PRUNING_KEEP_BLOCKS: Long = 6307200  // ~1 year

    // Fee model parameters (x/activity params)
    const val FEE_THRESHOLD_MULTIPLIER: Long = 3
    const val BASE_ACTIVITY_FEE: Long = 10000000000L         // 1 KKOT in uppyeo
    const val FEE_EXPONENT: Long = 5000                       // basis points (0.5)
    const val MAX_ACTIVITY_FEE: Long = 1000000000000L         // 100 KKOT in uppyeo
    const val MIN_FEEGRANT_QUOTA: Long = 8
    const val QUOTA_REDUCTION_RATE: Long = 5000               // basis points (0.5)
    const val FEEGRANT_FEE_EXEMPT: Boolean = true

    // Dual reward pool parameters (x/activity params)
    const val D_MIN: Long = 3000                   // basis points (0.3)
    const val FEE_TO_ACTIVITY_POOL_RATIO: Long = 8000  // basis points (0.8)

    // Node registration parameters (x/node params)
    const val MAX_REGISTRATIONS_PER_BLOCK: Long = 5
    const val REGISTRATION_COOLDOWN_BLKS: Long = 100
    const val REGISTRATION_DEPOSIT: String = "0"   // uppyeo
    const val MAX_TAGS: Long = 10
    const val MAX_TAG_LENGTH: Long = 32

    // Agent permission parameters
    val AGENT_ALLOWED_MSG_TYPES: List<String> = listOf(
        "/seocheon.activity.v1.MsgSubmitActivity",
        "/cosmos.bank.v1beta1.MsgSend",
    )
    val AGENT_FEEGRANT_ALLOWED_MSG_TYPES: List<String> = listOf(
        "/seocheon.activity.v1.MsgSubmitActivity",
    )

    const val AGENT_ADDRESS_CHANGE_COOLDOWN: Long = 17280  // 1 epoch

    // Time-block conversion (5s/block)
    const val BLOCKS_PER_HOUR: Long = 720
    const val BLOCKS_PER_DAY: Long = 17280
    const val BLOCKS_PER_YEAR: Long = 6307200
    const val UNBONDING_PERIOD_BLOCKS: Long = 362880   // ~21 days
    const val FEEGRANT_EXPIRY_BLOCKS: Long = 3110400   // ~180 days

    // Token denomination constants (6-stage system)
    // uppyeo(0) -> sal(2) -> pi(4) -> sum(6) -> hon(8) -> kkot(10)
    const val TOKEN_BASE_DENOM: String = "uppyeo"    // base denomination (10^0)
    const val TOKEN_SAL_DENOM: String = "sal"        // sal denomination (10^2)
    const val TOKEN_PI_DENOM: String = "pi"          // pi denomination (10^4)
    const val TOKEN_SUM_DENOM: String = "sum"        // sum denomination (10^6)
    const val TOKEN_HON_DENOM: String = "hon"        // hon denomination (10^8)
    const val TOKEN_DISPLAY_DENOM: String = "kkot"   // display denomination (10^10)

    // Denomination conversion factors (base unit: uppyeo)
    const val UPPYEO_PER_SAL: Long = 100
    const val UPPYEO_PER_PI: Long = 10000
    const val UPPYEO_PER_SUM: Long = 1000000
    const val UPPYEO_PER_HON: Long = 100000000
    const val UPPYEO_PER_KKOT: Long = 10000000000L

    // Gas defaults
    const val GAS_SUBMIT_ACTIVITY: Long = 200000
    const val GAS_WITHDRAW: Long = 300000
    const val GAS_SEND: Long = 100000
    const val GAS_FALLBACK: Long = 200000
    const val GAS_PRICE: Long = 250

    // Fee denomination
    const val FEE_DENOM: String = "uppyeo"

    // Default SDK configuration values
    const val DEFAULT_GAS_PRICE: String = "250uppyeo"
    const val DEFAULT_GAS_ADJUSTMENT: Double = 1.3
    const val DEFAULT_BROADCAST_MODE: String = "sync"
    const val DEFAULT_CONFIRM_TIMEOUT_MS: Long = 30000
    const val DEFAULT_CONFIRM_POLL_MS: Long = 1000

    // BIP44 path
    const val BIP44_COIN_TYPE: Int = 118
    const val BIP44_PATH: String = "m/44'/118'/0'/0/0"

    // Address prefix
    const val ADDRESS_PREFIX: String = "seocheon"
}
